add torchcodec to pytorch builds

lamikr opened the pull request:
Add support for checking out and building torchcodec during the pytorch build phase
- version of torchcodec depends from the pytorch version and torchcodecs/README.md lists which torchcodec and pytorch versions are compatible with each others. Because pytorch/.github/ci_commit_pins directory does not mention the torchcodec, we will check the pytorch version in pytorch_torchcodec_repo.py and then use that information to determine which torchcodec version to checkout.
- torchcodec source code checkout and build has been enabled in the linux CI job
- pytorch_torchcodec_repo.py will first install a torchcodec compatible version of setuptools before launching the torchcodec build command (setuptools > 0.80)

Todo:
- Verify windows builds
- Add smoketests

fixes: https://github.com/ROCm/TheRock/issues/1490

lamikr committed:
Add torchcodec to pytorch builds

- add support for checking out and building torchcodec
  as a part of the pytorch build process
- enable the torchcodec source code checkout and build
  in the linux CI job

Note:
- Version of torchcodec we checkout and build depends from the pytorch
  version and torchcodecs/README.md lists which torchcodec and pytorch
  versions are compatible with each others.
- Pinmapping file used by many other pytorch related subprojects does
  not yet exist for the torchcodec in directory pytorch/.github/ci_commit_pins
  and therefore we will check the pytorch version in
  pytorch_torchcodec_repo.py and then use that information to
  determine which torchcodec version to checkout.
- At least for now we will only support the torchcodec with pytorch 2.9
  and never releases and nightlies.
- Different pytorch components may require a different versions of
  python setuptools. Torchcodec requires that setup tools version needs to
  be never than 80.0. Because of that the build scripts install
  install a torchcodec compatible version of setuptools before launching the
  torchcodec build itself.

Todo:
- Add smoketests
- Add windows build support

fixes: https://github.com/ROCm/TheRock/issues/1490

Signed-off-by: Mika Laitio <mika.laitio@amd.com>

lamikr commented:
I am now trying to test the images build by ci with gfx1100 on Ubuntu 24.04. I used the rocm-sdk from 20260108 for building and install is still little tricky as I need to pickup packages from v2 and v2-staging dirs:

```
pip install --index-url https://rocm.devreleases.amd.com/v2/gfx110X-all rocm-sdk==7.11.0a20260108
pip install --index-url https://rocm.devreleases.amd.com/v2/gfx110X-all rocm-sdk-libraries-gfx110X-all==7.11.0a20260108
pip install --index-url https://rocm.devreleases.amd.com/v2/gfx110X-all rocm-sdk-devel==7.11.0a20260108
pip install --index-url https://rocm.devreleases.amd.com/v2-staging/gfx110X-all torch==2.11.0a0+rocm7.11.0a2026010
pip install --index-url https://rocm.devreleases.amd.com/v2-staging/gfx110X-all torchaudio==2.11.0a0+rocm7.11.0a20260108
pip install --index-url https://rocm.devreleases.amd.com/v2-staging/gfx110X-all torchcodec

```

Then also:

`apt install ffmpeg
`
At the moment also the LD_LIBRARY_PATH needs to be adjusted for now and I am still working on with that, so far I have set:

```
export LD_LIBRARY_PATH=~/therock_gfx1100/.venv_sdk2/lib/python3.12/site-packages/torch/lib
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:~/therock_gfx1100/.venv_sdk2/lib/python3.12/site-packages/_rocm_sdk_libraries_gfx110X_all/lib
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:~/therock_gfx1100/.venv_sdk2/lib/python3.12/site-packages/_rocm_sdk_core/lib
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:~/therock_gfx1100/.venv_sdk2/lib/python3.12/site-packages/_rocm_sdk_core/lib/host-math/lib
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:~/therock_gfx1100/.venv_sdk2/lib/python3.12/site-packages/_rocm_sdk_core/lib/rocm_sysdeps/lib

```

But I still had some error for loading ~/therock_gfx1100/.venv_sdk2/lib/python3.12/site-packages/torchcodec/libtorchcodec_core6.so

even thought

`ldd ~/therock_gfx1100/.venv_sdk2/lib/python3.12/site-packages/torchcodec/libtorchcodec_core6.so
`
looks fine

jeffdaily reviewed (changes requested):
Same changes requested as @amd-shiraz .

lamikr committed:
fix printouts based on to review

Signed-off-by: Mika Laitio <mika.laitio@amd.com>

lamikr committed:
build torchcodec v0.10.0 for pytorch 2.10

- update from v0.10.0-rc1 to v0.10.0

Signed-off-by: Mika Laitio <mika.laitio@amd.com>

lamikr committed:
remove the pip install of python build dependencies

I did local builds for pytorch 2.9, 2.10 and nightly
and was not able to reproduce anymore the error that
I have seen if not making sure before the torchcodec
build starts that the right pybind and setuptools
versions were installed.

It may have been a problem with some older torchcodec
version from the time I tested the building also
for pytorch 2.7 and pytorch 2.8

Signed-off-by: Mika Laitio <mika.laitio@amd.com>

lamikr committed:
Add pybind-dev debendencies back via pip install

Fixes following build error on ci-machine:

  By not providing "Findpybind11.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "pybind11",
  but CMake did not find one.

    Could not find a package configuration file provided by "pybind11" with any
    of the following names:
      pybind11Config.cmake
      pybind11-config.cmake

Signed-off-by: Mika Laitio <mika.laitio@amd.com>

ScottTodd commented on external-builds/pytorch/build_prod_wheels.py:
```diff
@@ -507,6 +510,16 @@ def do_build(args: argparse.Namespace):
     else:
         print("--- Not build pytorch-vision (no --pytorch-vision-dir)")
+    # Build pytorch torchcodec
+    if args.build_pytorch_torchcodec or (
```
nit: add a line break between the "Build pytorch vision" and the "Build pytorch torchcodec" sections

ScottTodd reviewed (changes requested):
The overall structure of the script and workflow changes looks fine. I'm wondering at a high level - is this working? I don't see any evidence of what testing has been done in the PR description and the code seems to be building with CUDA (and also ROCM?) disabled.

lamikr committed:
Update torchcodec github url

New upstream url:
https://github.com/meta-pytorch/torchcodec.git

Signed-off-by: Mika Laitio <mika.laitio@amd.com>

lamikr committed:
Apply argument parser suggestions from code review

Change audio related help texts to point to torchcodec.

Co-authored-by: Scott Todd <scott.todd0@gmail.com>

ScottTodd reviewed (commented):
Marking as 'reviewed' again to drop this from my queue. This PR has conflicts to resolve and we need to first answer if we need this at all, or if including the existing packages built upstream via https://github.com/ROCm/TheRock/blob/main/build_tools/third_party/s3_management/update_dependencies.py will be sufficient for our uses.

marbre commented:
> Is this still active? We are interested in a torchcodec build.

@lamikr ?
